@pelwell, after a rebase, I get a compile error on the dtsi:

```
Building branch unicam_clock
make[1]: Entering directory '/home/naushir/projects/linux_build/unicam_clock'
  GEN     Makefile
  DTC     arch/arm/boot/dts/bcm2708-rpi-b.dtb
/home/naushir/projects/linux/arch/arm/boot/dts/bcm270x.dtsi:86.22-100.5: ERROR (phandle_references): /soc/csi@7e800000: Reference to non-existent node or label "firmware_clocks"

/home/naushir/projects/linux/arch/arm/boot/dts/bcm270x.dtsi:102.22-116.5: ERROR (phandle_references): /soc/csi@7e801000: Reference to non-existent node or label "firmware_clocks"

  also defined at /home/naushir/projects/linux/arch/arm/boot/dts/bcm283x-rpi-csi1-2lane.dtsi:2.7-4.3
ERROR: Input tree has errors, aborting (use -f to force output)
make[2]: *** [scripts/Makefile.lib:286: arch/arm/boot/dts/bcm2708-rpi-b.dtb] Error 2
make[1]: *** [/home/naushir/projects/linux/Makefile:1249: dtbs] Error 2
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/naushir/projects/linux_build/unicam_clock'
make: *** [Makefile:179: sub-make] Error 2

```
I tried adding the following block to the end of the file (which is probably incorrect), but no luck:

```
&firmware {
	firmware_clocks: clocks {
		compatible = "raspberrypi,firmware-clocks";
		#clock-cells = <1>;
	};
};
```

My device tree knowledge is very limited, any idea what I should be doing go get firmware_clock recognised again?

--- a/Documentation/devicetree/bindings/media/bcm2835-unicam.txt
+++ b/Documentation/devicetree/bindings/media/bcm2835-unicam.txt
@@ -20,7 +20,7 @@ Required properties:
 - interrupts	: should contain the IRQ line for this Unicam instance.
 - clocks	: list of clock specifiers, corresponding to entries in
 		  clock-names property.
-- clock-names	: must contain an "lp" entry, matching entries in the
+- clock-names	: must contain "lp" and "vpu" entries, matching entries in the
 		  clocks property.
 
 Unicam supports a single port node. It should contain one 'port' child node
@@ -46,9 +46,9 @@ Example:
 		reg = <0x7e801000 0x800>,
 		      <0x7e802004 0x4>;
 		interrupts = <2 7>;
-		clocks = <&clocks BCM2835_CLOCK_CAM1>;
-		clock-names = "lp";
-
+		clocks = <&clocks BCM2835_CLOCK_CAM1>,
+			 <&firmware_clocks 4>;
+		clock-names = "lp", "vpu";
 		port {
 			csi1_ep: endpoint {
 				remote-endpoint = <&tc358743_0>;

--- a/arch/arm/boot/dts/bcm270x.dtsi
+++ b/arch/arm/boot/dts/bcm270x.dtsi
@@ -88,8 +88,9 @@
 			reg = <0x7e800000 0x800>,
 			      <0x7e802000 0x4>;
 			interrupts = <2 6>;
-			clocks = <&clocks BCM2835_CLOCK_CAM0>;
-			clock-names = "lp";
+			clocks = <&clocks BCM2835_CLOCK_CAM0>,
+				 <&firmware_clocks 4>;
+			clock-names = "lp", "vpu";
 			power-domains = <&power RPI_POWER_DOMAIN_UNICAM0>;
 			#address-cells = <1>;
 			#size-cells = <0>;
@@ -102,8 +103,9 @@
 			reg = <0x7e801000 0x800>,
 			      <0x7e802004 0x4>;
 			interrupts = <2 7>;
-			clocks = <&clocks BCM2835_CLOCK_CAM1>;
-			clock-names = "lp";
+			clocks = <&clocks BCM2835_CLOCK_CAM1>,
+				 <&firmware_clocks 4>;
+			clock-names = "lp", "vpu";
 			power-domains = <&power RPI_POWER_DOMAIN_UNICAM1>;
 			#address-cells = <1>;
 			#size-cells = <0>;

--- a/drivers/media/platform/bcm2835/bcm2835-unicam.c
+++ b/drivers/media/platform/bcm2835/bcm2835-unicam.c
@@ -89,6 +89,11 @@ MODULE_PARM_DESC(debug, "Debug level 0-3");
 		v4l2_err(&(dev)->v4l2_dev, fmt, ##arg)
 
 /*
+ * Unicam must request a minimum of 250Mhz from the VPU clock.
+ * Otherwise the input FIFOs overrun and cause image corruption.
+ */
+#define MIN_VPU_CLOCK_RATE (250 * 1000 * 1000)
+/*
  * To protect against a dodgy sensor driver never returning an error from
  * enum_mbus_code, set a maximum index value to be used.
  */
@@ -417,8 +422,10 @@ struct unicam_device {
 	void __iomem *base;
 	/* clock gating base address */
 	void __iomem *clk_gate_base;
-	/* clock handle */
+	/* lp clock handle */
 	struct clk *clock;
+	/* vpu clock handle */
+	struct clk *vpu_clock;
 	/* V4l2 device */
 	struct v4l2_device v4l2_dev;
 	struct media_device mdev;
@@ -1674,16 +1681,28 @@ static int unicam_start_streaming(struct vb2_queue *vq, unsigned int count)
 	unicam_dbg(1, dev, "Running with %u data lanes\n",
 		   dev->active_data_lanes);
 
+	ret = clk_set_min_rate(dev->vpu_clock, MIN_VPU_CLOCK_RATE);
+	if (ret) {
+		unicam_err(dev, "failed to set up VPU clock\n");
+		goto err_pm_put;
+	}
+
+	ret = clk_prepare_enable(dev->vpu_clock);
+	if (ret) {
+		unicam_err(dev, "Failed to enable VPU clock: %d\n", ret);
+		goto err_pm_put;
+	}
+
 	ret = clk_set_rate(dev->clock, 100 * 1000 * 1000);
 	if (ret) {
-		unicam_err(dev, "failed to set up clock\n");
-		goto err_pm_put;
+		unicam_err(dev, "failed to set up CSI clock\n");
+		goto err_vpu_clock;
 	}
 
 	ret = clk_prepare_enable(dev->clock);
 	if (ret) {
 		unicam_err(dev, "Failed to enable CSI clock: %d\n", ret);
-		goto err_pm_put;
+		goto err_vpu_clock;
 	}
 
 	for (i = 0; i < ARRAY_SIZE(dev->node); i++) {
@@ -1717,6 +1736,11 @@ static int unicam_start_streaming(struct vb2_queue *vq, unsigned int count)
 err_disable_unicam:
 	unicam_disable(dev);
 	clk_disable_unprepare(dev->clock);
+err_vpu_clock:
+	ret = clk_set_min_rate(dev->vpu_clock, 0);
+	if (ret)
+		unicam_err(dev, "failed to reset the VPU clock\n");
+	clk_disable_unprepare(dev->vpu_clock);
 err_pm_put:
 	unicam_runtime_put(dev);
 err_streaming:
@@ -1734,6 +1758,8 @@ static void unicam_stop_streaming(struct vb2_queue *vq)
 	node->streaming = false;
 
 	if (node->pad_id == IMAGE_PAD) {
+		int ret;
+
 		/*
 		 * Stop streaming the sensor and disable the peripheral.
 		 * We cannot continue streaming embedded data with the
@@ -1743,6 +1769,12 @@ static void unicam_stop_streaming(struct vb2_queue *vq)
 			unicam_err(dev, "stream off failed in subdev\n");
 
 		unicam_disable(dev);
+
+		ret = clk_set_min_rate(dev->vpu_clock, 0);
+		if (ret)
+			unicam_err(dev, "failed to reset the min VPU clock\n");
+
+		clk_disable_unprepare(dev->vpu_clock);
 		clk_disable_unprepare(dev->clock);
 		unicam_runtime_put(dev);
 
@@ -2742,8 +2774,15 @@ static int unicam_probe(struct platform_device *pdev)
 
 	unicam->clock = devm_clk_get(&pdev->dev, "lp");
 	if (IS_ERR(unicam->clock)) {
-		unicam_err(unicam, "Failed to get clock\n");
+		unicam_err(unicam, "Failed to get lp clock\n");
 		ret = PTR_ERR(unicam->clock);
+		goto err_unicam_put;
+	}
+
+	unicam->vpu_clock = devm_clk_get(&pdev->dev, "vpu");
+	if (IS_ERR(unicam->vpu_clock)) {
+		unicam_err(unicam, "Failed to get vpu clock\n");
+		ret = PTR_ERR(unicam->vpu_clock);
 		goto err_unicam_put;
 	}
 

